chain/chain/src/chain.rs

```diff
@@ -3145,6 +3179,34 @@ impl Chain {
         apply_result?;
 
         let shard_state_header = self.get_state_header(shard_id, sync_hash)?;
+        let chunk = shard_state_header.cloned_chunk();
+
+        let block_hash = chunk.prev_block();
+        let block_height = self.get_block_header(block_hash)?.height();
+
+        // Flat storage must not exist at this point because leftover keys corrupt its state.
+        assert!(self.runtime_adapter.get_flat_storage_state_for_shard(shard_id).is_none());
+
+        // We synced shard state on top of _previous_ block for chunk in shard state header and applied state parts to
+        // flat storage. Now we can set flat head to hash of this block and create flat storage.
+        #[cfg(feature = "protocol_feature_flat_state")]
+        {
+            let mut store_update = self.runtime_adapter.store().store_update();
+            store_helper::set_flat_head(&mut store_update, shard_id, block_hash);
+            store_update.commit()?;
+        }
+
+        match self.runtime_adapter.try_create_flat_storage_state_for_shard(
```

Comment:
@Longarithm @mm-near I think this doesn't create the FlatStorage object, it only creates the FlatStorageState if the status is already `FlatStorageStateStatus::Ready`. I don't see where in the logic before this would update the status to be ready, so I assume the creation of FlatState will be done later through the call of `update_flat_storage_for_block` which calls `update_status`, which triggers the creation of FlatState asynchronously in other threads. Every time when `update_status` is called, it checks the response from the other threads and update `FlatStorageStateStatus` accordingly. That's why I said when state sync is finished, the creation of FlatStorage is not necessarily finished. Please let me know if my understanding is wrong.